src/execution/execute.ts

```diff
@@ -790,106 +791,128 @@ function completeLeafValue(
+  possibleRuntimeTypeName: PromiseOrValue<string | undefined>,
   fieldNodes: ReadonlyArray<FieldNode>,
   info: GraphQLResolveInfo,
+  path: Path,
   result: unknown,
-): GraphQLObjectType {
-  if (runtimeTypeName == null) {
+): PromiseOrValue<ObjMap<unknown>> {
+  if (isPromise(possibleRuntimeTypeName)) {
+    return possibleRuntimeTypeName.then((resolved) =>
+      completeAbstractValueImpl(
+        exeContext,
+        abstractType,
+        resolved,
+        fieldNodes,
+        info,
+        path,
+        result,
+      ),
+    );
+  }
+
+  if (possibleRuntimeTypeName == null) {
     throw new GraphQLError(
-      `Abstract type "${returnType.name}" must resolve to an Object type at runtime for field "${info.parentType.name}.${info.fieldName}". Either the "${returnType.name}" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.`,
+      `Abstract type "${abstractType.name}" must resolve to an Object type or an intermediate Interface type at runtime for field "${info.parentType.name}.${info.fieldName}". Either the "${abstractType.name}" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.`,
       { nodes: fieldNodes },
     );
   }
 
   // releases before 16.0.0 supported returning `GraphQLObjectType` from `resolveType`
   // TODO: remove in 17.0.0 release
-  if (isObjectType(runtimeTypeName)) {
+  if (isObjectType(possibleRuntimeTypeName)) {
     throw new GraphQLError(
       'Support for returning GraphQLObjectType from resolveType was removed in graphql-js@16.0.0 please return type name instead.',
     );
   }
 
-  if (typeof runtimeTypeName !== 'string') {
+  if (typeof possibleRuntimeTypeName !== 'string') {
     throw new GraphQLError(
-      `Abstract type "${returnType.name}" must resolve to an Object type at runtime for field "${info.parentType.name}.${info.fieldName}" with ` +
-        `value ${inspect(result)}, received "${inspect(runtimeTypeName)}".`,
+      `Abstract type "${abstractType.name}" must resolve to an Object type or an intermediate Interface type at runtime for field "${info.parentType.name}.${info.fieldName}" with ` +
+        `value ${inspect(result)}, received "${inspect(
+          possibleRuntimeTypeName,
+        )}".`,
     );
   }
 
-  const runtimeType = exeContext.schema.getType(runtimeTypeName);
-  if (runtimeType == null) {
+  const possibleRuntimeType = exeContext.schema.getType(
+    possibleRuntimeTypeName,
+  );
+  if (possibleRuntimeType == null) {
     throw new GraphQLError(
-      `Abstract type "${returnType.name}" was resolved to a type "${runtimeTypeName}" that does not exist inside the schema.`,
+      `Abstract type "${abstractType.name}" was resolved to a type "${possibleRuntimeTypeName}" that does not exist inside the schema.`,
       { nodes: fieldNodes },
     );
   }
 
-  if (!isObjectType(runtimeType)) {
+  if (isInterfaceType(possibleRuntimeType)) {
```

Comment:
what about union types?

Reply:
Yes, if unions ever implement interfaces, they could potentially be a link in the chain!